index.js

```diff
@@ -1,30 +1,46 @@
 import fp from 'fastify-plugin'
 import onFinished from 'on-finished'
-import defaults from 'defaults'
 
 import { DEFAULT_OPTIONS, HEADERS } from './lib/constants.js'
 import { Store } from './lib/store.js'
 import { convertToMs, calculateDelay } from './lib/helpers.js'
 
 const slowDownPlugin = async (fastify, settings) => {
-  const options = defaults(settings, DEFAULT_OPTIONS)
+  const options = { ...DEFAULT_OPTIONS, ...settings }
   const store = new Store(convertToMs(options.timeWindow))
 
+  if (!fastify.hasRequestDecorator('slowDown')) {
+    fastify.decorateRequest('slowDown', {
+      limit: options.delayAfter
+    })
+  }
+
   fastify.addHook('onClose', () => store.close())
 
   fastify.addHook('onRequest', async (req, reply) => {
     const requestCounter = store.incrementOnKey(options.keyGenerator(req))
     const delayMs = calculateDelay(requestCounter, options)
+    const remainingRequests = Math.max(options.delayAfter - requestCounter, 0)
 
     reply.header(HEADERS.limit, options.delayAfter)
-    reply.header(
-      HEADERS.remaining,
-      Math.max(options.delayAfter - requestCounter, 0)
-    )
+    reply.header(HEADERS.remaining, remainingRequests)
+
+    if (fastify.hasRequestDecorator('slowDown')) {
+      req.slowDown = {
+        ...req.slowDown,
+        indexRequest: requestCounter,
+        remainingRequests
+      }
+    }
 
     if (delayMs <= 0) {
       return
     }
+
+    if (fastify.hasRequestDecorator('slowDown')) {
```

Comment:
You can avoid to check the decorator presence twice, extracting a dedicated variable.

Additionally, you can check it once and move the following statement on line 33:
```
req.slowDown.delay = delayMs <= 0 ? undefined : delayMs
```